### PR TITLE
Keep selection and don't overwrite user entered text in the url bar when opening new tabs

### PR DESCRIPTION
**Pull request by victorporof**
Fixes https://github.com/mozilla/tofino/issues/1177 

Signed-off-by: Victor Porof <vporof@mozilla.com>

### CODE REVIEW

File: app/ui/browser-modern/sagas/ui-sagas.js [hunk 1]
```diff
@@ -32,13 +33,35 @@ export default function() {
...
+  const interaction = urlbar.dataset.interaction;
+  if (interaction === 'user-is-typing') {
+    logger.warn(`Skipped setting \`value\` on the urlbar because: ${interaction}.`);
+    urlbar.defaultValue = value;
+    return;
+  }
+
+  const { selectionStart, selectionEnd } = urlbar;
+  const allTextSelected = selectionStart === 0 && selectionEnd === urlbar.value.length;
+  const isActiveElement = document.activeElement === urlbar;
```
Comment (jsantell) [inline app/ui/browser-modern/sagas/ui-sagas.js]: Maybe pass in `document` for tests

File: app/ui/browser-modern/sagas/ui-sagas.js [hunk 2]
```diff
@@ -32,13 +33,35 @@ export default function() {
...
 
-function* setURLBarValue({ urlbar, value }) {
+function* setURLBarValue({ urlbar, value, options = { keepSelection: true } }) {
+  // When the user is interacting with the url bar, don't just change all the
+  // displayed text directly, only the default value, to prevent replacing
+  // user entered text, but allowing the intended text to be displayed when
+  // cancelling input (e.g. by pressing 'Escape').
+  const interaction = urlbar.dataset.interaction;
+  if (interaction === 'user-is-typing') {
+    logger.warn(`Skipped setting \`value\` on the urlbar because: ${interaction}.`);
```
Comment (jsantell) [inline app/ui/browser-modern/sagas/ui-sagas.js]: Will this get noisy?

File: app/ui/shared/widgets/search.jsx [hunk 3]
```diff
@@ -62,9 +72,20 @@ class Search extends Component {
...
 
-  handleClick = () => {
+  handleClick = e => {
     this.select();
-    this.props.onClick();
+    this.props.onClick(e);
+  }
+
+  handleKeyDown = e => {
+    this.setState({ interaction: 'user-is-typing' });
```
Comment (jsantell) [inline app/ui/shared/widgets/search.jsx]: I think consts for the interaction types would be good here -- elsewhere, like sagas, can remain primitives

File: app/ui/shared/widgets/search.jsx [hunk 4]
```diff
@@ -62,9 +72,20 @@ class Search extends Component {
...
   }
 
-  handleClick = () => {
+  handleClick = e => {
     this.select();
-    this.props.onClick();
+    this.props.onClick(e);
+  }
+
+  handleKeyDown = e => {
```
Comment (jsantell) [inline app/ui/shared/widgets/search.jsx]: I think turning this into a debounced handler via lodash would be cleaner

**Comment by victorporof** [comment]
Agreed with all comments, will update.

File: app/ui/browser-modern/sagas/ui-sagas.js (hunk 2 above)
Comment (victorporof) [inline app/ui/browser-modern/sagas/ui-sagas.js]: Not if you're constantly typing :)

**Comment by victorporof** [comment]
Addressed comments.

**Pull request merged**